### PR TITLE
Add immutable tag in project detail

### PR DESCRIPTION
**Pull request by jwangyangls**
Signed-off-by: Yogi_Wang <yawang@vmware.com>

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -32,7 +32,7 @@ src/portal/typings/
 
 src/portal/src/**/*.js
 src/portal/src/**/*.js.map
-
+src/portal/lib/coverage
```
Comment (AllForNothing) [inline .gitignore]: I've also modified here in PR #9361

File: src/portal/src/app/harbor-routing.module.ts [hunk 2]
```diff
@@ -276,6 +277,16 @@ const harborRoutes: Routes = [
             },
             component: TagRetentionComponent
           },
+          {
+            path: 'immutable-tag',
+            data: {
+              permissionParam: {
+                resource: USERSTATICPERMISSION.TAG_RETENTION.KEY,
```
Comment (AllForNothing) [inline src/portal/src/app/harbor-routing.module.ts]: Use same resource  key as tag-retantion?

File: src/portal/src/app/project/immutable-tag/add-rule/add-rule.component.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    canNotAdd(): boolean {
+        if (this.onGoing) {
+            return true;
+        }
+        if (!this.isAdd && compareValue(this.editRuleOrigin, this.rule)) {
+            return true;
+        }
+        return !(
+            this.rule.scope_selectors.repository[0].pattern
```
Comment (zhoumeina) [inline src/portal/src/app/project/immutable-tag/add-rule/add-rule.component.ts]: How do you know this.rule.scope_selectors is not null

File: src/portal/src/app/project/immutable-tag/add-rule/add-rule.component.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    cancel() {
+        this.close();
+    }
+
+    add() {
+        // remove whitespaces
+        this.rule.scope_selectors.repository[0].pattern = this.rule.scope_selectors.repository[0].pattern.replace(/\s+/g, "");
+        this.rule.tag_selectors[0].pattern = this.rule.tag_selectors[0].pattern.replace(/\s+/g, "");
```
Comment (zhoumeina) [inline src/portal/src/app/project/immutable-tag/add-rule/add-rule.component.ts]: show add validator to check if scope_selectors. 
For repository, it is the same.

**Pull request merged**